src/jdk.incubator.code/share/classes/jdk/incubator/code/type/ClassType.java

```diff
@@ -83,35 +83,27 @@ private static ParameterizedType makeReflectiveParameterizedType(Class<?> base,
         return sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl.make(base, typeArgs, owner);
     }
 
-    @Override
-    public ExternalizedTypeElement externalize() {
-        List<ExternalizedTypeElement> args = typeArguments.stream()
-                .map(TypeElement::externalize)
-                .toList();
-
-        ExternalizedTypeElement td = new ExternalizedTypeElement(toClassName(), args);
-        if (enclosing != null) {
-            td = new ExternalizedTypeElement(".", List.of(enclosing.externalize(), td));
-        }
-        return td;
-    }
-
     @Override
     public String toString() {
         String prefix = enclosing != null ?
-                enclosing + "$":
+                enclosing + "::":
                 (!type.packageName().isEmpty() ?
                         type.packageName() + "." : "");
         String name = enclosing == null ?
                 type.displayName() :
-                type.displayName().substring(enclosing.type.displayName().length() + 1);
+                escape(type.displayName().substring(enclosing.type.displayName().length() + 1));
         String typeArgs = hasTypeArguments() ?
                 typeArguments().stream().map(JavaType::toString)
                         .collect(Collectors.joining(", ", "<", ">")) :
                 "";
         return String.format("%s%s%s", prefix, name, typeArgs);
     }
 
+    static String escape(String s) {
```

Comment:
This is interesting (see also `ExternalizedType::toString`): if the type identifier is not a valid Java identifier, we need to wrap it in quotes -- otherwise we might have parsing issues when doing a round-trip. This comes up e.g. in local class names, which start with a number.